utils/patientmatching/rule_based_patient_matcher.bal

```diff
@@ -0,0 +1,274 @@
+public type RulesRecord record {|
+    # array of fhirpath rules
+    string[] fhirpathArray;
+|};
+
+# Record to hold MPI database configuration details.
+public type MPIDbConfig record {|
+    # host name of MPI database
+    string host;
+    # port of MPI database
+    int port;
+    # username of MPI database
+    string username;
+    # password of MPI database
+    string password;
+    # database name of MPI database
+    string database;
+|};
+
+# Implementation of the RuleBasedPatientMatching Algorithm.
+public isolated class RuleBasedPatientMatcher {
+    *PatientMatcher;
+
+    public isolated function matchPatients(PatientMatchRequest patientMatchRequest, ConfigurationRecord? config) returns error|http:Response {
+        json[] parametersArray = patientMatchRequest.'parameter;
+        r4:Patient|error sourcePatient = (check parametersArray[0].'resource).cloneWithType();
+        string|error strPatientCount = (check parametersArray[1].valueInteger).cloneWithType();
+        string|error strOnlyCertainMatches = (check parametersArray[2].valueBoolean).cloneWithType();
+        if sourcePatient is error {
+            return throwFHIRError("Error occurred while getting the source patient from the request.", cause = sourcePatient);
+        }
+        if strPatientCount is error {
+            return throwFHIRError("Error occurred while getting the expected patient count from the request." , cause = strPatientCount);
+        }
+        if strOnlyCertainMatches is error {
+            return throwFHIRError("Error occurred while getting the OnlyCertainMatches boolean flag from the request.",cause = strOnlyCertainMatches);
+        }
+
+        int|error patientCount = int:fromString(strPatientCount);
+        boolean|error onlyCertainMatches = boolean:fromString(strOnlyCertainMatches);
+        if patientCount is error {
+            return throwFHIRError("Error occurred while casting the patient count string to integer." ,cause = patientCount);
+        }
+        if onlyCertainMatches is error {
+            return throwFHIRError("Error occurred while casting the onlyCertainMatches flag from string to boolean.",cause = onlyCertainMatches);
+        }
+
+        r4:BundleEntry[]|error? patientArray = self.getMatchingPatients(<r4:Patient>sourcePatient, config ?: {});
+        if patientArray is () {
+            http:Response response = new;
+            response.setJsonPayload("No matching patient found");
+            return response;
+        }
+        if patientArray is error {
+            return throwFHIRError("Error occurred while getting the matching patients from MPI.", cause = patientArray);
+        }
+        http:Response response = new;
+        if onlyCertainMatches is true && patientArray.length() > 1 {
+            response.setJsonPayload("Multiple matching patients found while onlyCertainMatches flag is true");
+        }
+        if patientArray.length() < patientCount {
+            r4:Bundle bundle = {
```

Comment:
Do we need to define this as `r4:Bundle`?

Reply:
It is defined in the FHIR Specification. So we need to follow that.